waiter/src/waiter/token.clj

```diff
@@ -719,3 +725,119 @@
                                              :status http-405-method-not-allowed})))
     (catch Exception ex
       (utils/exception->response ex req))))
+
+(defn make-index-event
+  [type object]
+  {:object object :type type})
+
+(defn start-tokens-watch-maintainer
+  [kv-store tokens-update-chan tokens-watch-channels-update-chan watch-refresh-timeout-ms exit-chan]
+  (let [tokens-event-chan (au/latest-chan)
+        tokens-event-mult (async/mult tokens-event-chan)
+        query-chan (async/chan)
+        get-token->token-index-fn (fn []
+                                    (->> kv-store
+                                         list-token-owners
+                                         (reduce
+                                           (fn [current-token-index-map owner]
+                                             (reduce
+                                               (fn [inner-token-index-map [token entry]]
+                                                 (->> (assoc entry :owner owner :token token)
+                                                      (assoc inner-token-index-map token)))
+                                               current-token-index-map
+                                               (list-index-entries-for-owner kv-store owner :refresh true)))
+                                           {})))
+        state-atom (atom {:token->token-index (get-token->token-index-fn)
+                          :watches-count 0})
+        go-chan
+        (async/go
+          (try
+            (loop [{:keys [token->token-index watches-count] :as current-state} @state-atom
+                   watch-refresh-timeout-chan (async/timeout watch-refresh-timeout-ms)]
```

Comment:
The way we are using `watch-refresh-timeout-chan`, I would consider using `async-utils/timer-chan`.
Also, this channel is a good candidate to pass in as an argument as it can simplify the unit tests and avoid the `Thread/sleep` calls.
In the unit tests we can pass in a regular channel as the `watch-refresh-timeout-chan` and write data to it to trigger processing of the branch we want.